### PR TITLE
feat: redefine sequencer blocks, celestia blobs as protobuf

### PR DESCRIPTION
**Pull request by SuperFluffy**
## Summary
This patch defines protobuf versions of various types that are currently only defined in Rust and exchange as json-serialized text. 

## Background
`astria-proto` is intended to be the single source of truth for the wire format of data that is exchanged between services. At the moment `SequencerBlockData`, `SequencerNamespaceData`, `RollupNamespaceData` and `InclusionProof` are only defined in Rust. Some older versions of these exist as protobuf but were never used in the public interfaces.

## Changes
- Replace `SequencerBlockData` by a protobuf `SequencerBlock` (rewriting this protobuf type is technically a breaking change, but it was never used, so in practice it is not)
- Replace `SequencerNamespaceData` and `RollupNamespaceData` by protobuf `CelestiaSequencerBlob` and `CelestiaRollupBlob`.
- Introduce other protobuf types like `RollupTransactions`, `Proof` to decode/encode all required objects
- Write idiomatic rust types for all generated protobuf representations to uphold invariants

## Testing
- all tests have been updated and still pass
- testing this is in a devnet to check that the data still flows is open

## Related Issues

https://github.com/astriaorg/astria/pull/12 originally defined sequencer blocks and transactions as protobuf, but these were never actually implemented.

https://github.com/astriaorg/astria/issues/86 is the reason this effort was started.

### CODE REVIEW

**Comment by SuperFluffy** [comment]
Opened a new PR because something broke https://github.com/astriaorg/astria/pull/355. No changes with respect to the most recent review.

File: crates/astria-proto/proto/astria/sequencer/v1alpha2/celestia.proto
Comment (noot) [inline crates/astria-proto/proto/astria/sequencer/v1alpha2/celestia.proto:21]:
```suggestion
    // `sequencer_block_header.data_hash` (to be found in `CelestiaSequencerData`).
```

**Review by noot — APPROVED** [review]
looks good! are you planning to do #20 (and move the types to astria-proto) in a follow up?

also is  #31 still relevant or can it be closed?

**Comment by SuperFluffy** [comment]
@noot yeah, #20 should be the immediate follow-up to this PR

#31 will probably be outdated with this change.

I have dismissed your review because I think this PR needs more discussion in light of #401 (to ensure this does not accidentally sneak in).

**Comment by noot** [comment]
we can put #401 on hold for now - don't think we need to change the verification stuff yet, probably going to close #401 and open another broader issue. so this should be good to go in!

**Review by noot — APPROVED** [review]

File: crates/astria-proto/proto/astria/sequencer/v1alpha1/block.proto
Comment (joroshiba) [inline crates/astria-proto/proto/astria/sequencer/v1alpha1/block.proto:1]: are these files supposed to be in v1alpha2? Where are the v1alpha2 generated files coming from?

Comment (SuperFluffy) [reply crates/astria-proto/proto/astria/sequencer/v1alpha1/block.proto:1]: forgot to push the cleanup

**Comment by SuperFluffy** [comment]
The proto lint will fail because this is a breaking proto buf change (which is ok, because the proto definitions were never used)

File: .github/workflows/test.yml
Comment (SuperFluffy) [inline .github/workflows/test.yml:91]: I set all optimism tests as `#[ignore]` because I did not have the required dependencies installed.

File: crates/astria-proto/proto/astria/sequencer/v1alpha1/transaction.proto
Comment (SuperFluffy) [inline crates/astria-proto/proto/astria/sequencer/v1alpha1/transaction.proto:57]: Renamed this field to be in line with renaming chain ID -> rollup ID.

This is not a breaking protobuf change (field names don't matter)

File: crates/astria-proto/src/lib.rs
Comment (SuperFluffy) [inline crates/astria-proto/src/lib.rs:1]: I did not want to worry about casting usize to u64 and back. Astria Merkle is written entirely using `usize` so it works on any architecture. However, moving in between architectures is more tricky. So 64 bit it is.

File: crates/astria-proto/proto/astria/sequencer/v1alpha1/block.proto
Comment (SuperFluffy) [inline crates/astria-proto/proto/astria/sequencer/v1alpha1/block.proto:26]: Rather than using a map here this is now a list of `(Id, [bytes])` tuples. This means that any language will get these in the correct order.

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (SuperFluffy) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:1]: Note on this file: we will split this in the future because it's getting too unwieldy. It should be done in a separate PR however to not mix feature changes and simply moving code around

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (SuperFluffy) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:1012]: @noot @joroshiba I wanted to especially flag this: I renamed chain ID to rollup ID everywhere.

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (SuperFluffy) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:1573]: This method used to be a free standing method in `astria-celestia-client`. Because all these types are defined here it's natural to leave this here.

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (SuperFluffy) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:1613]: This was moved from `astria-sequencer-types` and enriched by checking more invariants.

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (SuperFluffy) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:2391]: This was moved/adapted from astria-sequencer. Because other parts of the code are interacting with it (mainly through tests) it made sense to me to put it here, also because it's closely relatted to `derive_merkle_from_rollup_txes`

File: crates/astria-conductor/src/data_availability/block_verifier.rs
Comment (SuperFluffy) [inline crates/astria-conductor/src/data_availability/block_verifier.rs:449]: This check no longer made sense because the method to fetch the blobs in `astria-celestia-client` now also ensures that only blobs that are related to the sequencer blob are returned.

File: crates/astria-conductor/src/conductor.rs
Comment (joroshiba) [inline crates/astria-conductor/src/conductor.rs:110]: Nice

File: crates/astria-proto/proto/astria/sequencer/v1alpha1/celestia.proto
Comment (noot) [inline crates/astria-proto/proto/astria/sequencer/v1alpha1/celestia.proto:47]:
```suggestion
    // The proof that the rollup transactions are included in sequencer block header.
```

File: crates/astria-proto/proto/astria/sequencer/v1alpha1/celestia.proto
Comment (noot) [inline crates/astria-proto/proto/astria/sequencer/v1alpha1/celestia.proto:52]: no `rollup_ids_root`?

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (noot) [reply crates/astria-proto/src/native/sequencer/v1alpha1.rs:1]: linking #602

File: crates/astria-proto/proto/astria/sequencer/v1alpha1/celestia.proto
Comment (SuperFluffy) [reply crates/astria-proto/proto/astria/sequencer/v1alpha1/celestia.proto:52]: Can be reconstructed ad hoc from the list of IDs.

In fact one needs to do this always anyways to ensure the stored is correct. But then we can just skip it.

See the comment and the constructor of the idiomatic native type.

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (noot) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:1489]: how does `IndexMap`'s ordering work compared to `BTreeMap`? assuming it's different, that's a breaking change

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (noot) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:1492]: what is `MTH`? merkle tree hash?

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (noot) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:1501]: same question as on the proto - how come the proofs are here but the commitments aren't? is it because you plan to always reconstruct the `rollup_transactions_root` from `rollup_transactions`?

File: crates/astria-proto/proto/astria/sequencer/v1alpha1/block.proto
Comment (noot) [reply crates/astria-proto/proto/astria/sequencer/v1alpha1/block.proto:42]: ok, i see you are planning to just reconstruct the commitments everywhere

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (SuperFluffy) [reply crates/astria-proto/src/native/sequencer/v1alpha1.rs:1489]: This new grouping mechanism is used inside sequencer as well. You can see that the snapshot test is unaffected.

When the indexmap is constructed it is ordered by its keys after the fact. But this is the same order as that used for the btreemap.

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (SuperFluffy) [reply crates/astria-proto/src/native/sequencer/v1alpha1.rs:1501]: Yeah, but this was always done anyway to ensure these data types uphold invariants and that the stored root and the data it is constructed from is correct.

But it need not be written to the wire format.

If it ever becomes necessary to give repeated access to this root we can add it to the idiomatic types as is done for the cometbft header block hash.

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (noot) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:1702]:
```suggestion
    /// Converts from the raw decoded protobuf representation of this type.
```

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (noot) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:1725]: why the `'proof` syntax over just returning the error?

Comment (SuperFluffy) [reply crates/astria-proto/src/native/sequencer/v1alpha1.rs:1725]: There are two steps in this block: one ensures the protobuf value was set, the next one actually does the conversion.

These two steps belong together logically and I didn't want to declare variables at the function level namespace.

I also think this is neat.

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (noot) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:2391]:
```suggestion
pub fn group_sequence_actions_in_signed_transactions_by_rollup_id(
```

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (joroshiba) [inline crates/astria-proto/src/native/sequencer/v1alpha1.rs:1840]: Why the "unchecked" verbiage here? There is nothing checked when converting? It's just pub access vs not?

File: crates/astria-sequencer-types/src/tendermint.rs
Comment (noot) [inline crates/astria-sequencer-types/src/tendermint.rs:1]: delete this file?

File: crates/astria-conductor/src/conductor.rs
Comment (noot) [inline crates/astria-conductor/src/conductor.rs:99]: rename the config value also to `rollup_id`

File: crates/astria-proto/proto/astria/sequencer/v1alpha1/block.proto
Comment (SuperFluffy) [reply crates/astria-proto/proto/astria/sequencer/v1alpha1/block.proto:26]: The order is important for constructing the Merkle Hash Tree of the roll-up transactions.

The order in this sequencer block is the canonical order used by the sequencer.

Maps vs lists aside, I'm not even sure that we can rely on an order on bytes in other languages. So this is cleaner.

File: crates/astria-conductor/src/conductor.rs
Comment (noot) [inline crates/astria-conductor/src/conductor.rs:181]: yeah, do you wanna delete/update it?

File: crates/astria-celestia-client/src/client.rs
Comment (noot) [inline crates/astria-celestia-client/src/client.rs:270]:
```suggestion
/// + cannot decode/convert to [`CelestiaRollupBlob`]
```

File: crates/astria-celestia-client/src/client.rs
Comment (noot) [inline crates/astria-celestia-client/src/client.rs:272]:
```suggestion
/// + the proof, ID, and transactions recorded in the blob cannot be verified against the sequencer
```

**Review by noot — APPROVED** [review]

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (SuperFluffy) [reply crates/astria-proto/src/native/sequencer/v1alpha1.rs:1489]: BtreeMap is balancing while it's being constructed. That gives nlogn overhead. Insertion the indexmap is linear and the final ordering has the same complexity.

Access on the other hand is O(1) where it is log n for the btreemap.

The indexmap is cache efficient, the btreemap is not.

So for our purpose the index map is strictly better.

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (noot) [reply crates/astria-proto/src/native/sequencer/v1alpha1.rs:1725]:
```
break 'proof Err(SequencerBlockError::FieldNotSet(
   "rollup_transactions_proof",
));
```

this is logically the same as:

```
return Err(SequencerBlockError::FieldNotSet(
   "rollup_transactions_proof",
));
```

right?  just checking for my own sanity lol

File: crates/astria-proto/src/native/sequencer/v1alpha1.rs
Comment (SuperFluffy) [reply crates/astria-proto/src/native/sequencer/v1alpha1.rs:1840]: Yeah, thats the case here.

Honestly no reason other than being symmetric with the sequencer blob.

I'm not a big fan of the unchecked versions. But yes, it's to give pub access to the fields to avoid allocations.

**Review by joroshiba — APPROVED** [review]

File: crates/astria-conductor/src/conductor.rs
Comment (SuperFluffy) [reply crates/astria-conductor/src/conductor.rs:99]: I wasn't sure about this. It's m technically not the ID but just the name. The ID are the 32 bytes and I don't want to cause confusion between these two.

Is rollup_name good? Also CC @joroshiba

**Pull request merged**
